// app/models/index.js
"use 6to5";

const Rx = require('rx-dom');
const pipe = require('app/pipe');

module.exports = function model() {
  var heightOffsetChanged$ = new Rx.Subject();
  var changeHeightStarted$ = new Rx.Subject();

  var changeHeight$ = heightOffsetChanged$.map(offset => {
    return function(model) {
      return {
        height: model.height - model.offset + offset,
        offset: offset,
        isShow: model.isShow
      };
    };
  });

  var startChangeHeight$ = changeHeightStarted$.map(() => {
    return function(model) {
      return {
        height: model.height,
        offset: 0,
        isShow: model.isShow
      };
    };
  });

  var model$ = changeHeight$.merge(startChangeHeight$).startWith({
    height: 100,
    isShow: true,
    offset: 0
  }).scan(function(model, modification) {
    return modification(model);
  });

  return {
    model$: model$,
    observe(action) {
      pipe(action.heightOffsetChanged$, heightOffsetChanged$);
      pipe(action.changeHeightStarted$, changeHeightStarted$);
    }
  };
};
